proposals/NNNN-method-and-initializer-keypaths.md

```diff
@@ -0,0 +1,165 @@
+2. With argument application: The key path references the method with arguments already applied.
+
+Continuing our `Calculator` example, we can write either:
+
+```swift
+let squareWithoutArgs: KeyPath<Calculator, (Int) -> Int> = \Calculator.square
+let squareWithArgs: KeyPath<Calculator, Int> = \Calculator.square(of: 3)
+```
+
+If the member is a metatype (e.g., a static method, class method, initializer, or when referring to the type of an instance), you must explicitly include `.Type` in the key path root type.
+
+```swift
+struct Calculator {
+  func add(_ a: Int, _ b: Int) -> Int {
+    return a + b
+  }
+}
+
+let calc = Calculator.self
+let addKeyPath: KeyPath<Calculator.Type, (Calculator) -> (Int, Int) -> Int> = \Calculator.Type.add
+```
+
+Here, `addKeyPath` is a key path that references the add method of `Calculator` as a metatype member. The key path’s root type is `Calculator.Type`, and it resolves to an unapplied instance method: `(Calculator) -> (Int, Int) -> Int`. This represents a curried function where the first step binds an instance of `Calculator`, and the second step applies the method arguments.
+
+```swift
+let addFunction = calc[keyPath: addKeyPath]
+let fullyApplied = addFunction(Calculator())(20, 30)`
+```
+
+`addFunction` applies an instance of Calculator to the key path method. `fullyApplied` further applies the arguments (20, 30) to produce the final result.
+
+### Overloads
+
+Keypaths to methods with the same base name and distinct argument labels can be disambiguated by explicitly including the argument labels:
+
+```swift
+struct Calculator {
+  var subtract: (Int, Int) -> Int { return { $0 + $1 } }
+  func subtract(this: Int) -> Int { this + this}
+  func subtract(that: Int) -> Int { that + that }
+}
+  
+let kp1 = \S.subtract // KeyPath<S, (Int, Int) -> Int
+let kp2 = \S.subtract(this:) // WritableKeyPath<S, (Int) -> Int>
+let kp3 = \S.subtract(that:) // WritableKeyPath<S, (Int) -> Int>
+let kp4 = \S.subtract(that: 1) // WritableKeyPath<S, Int>
+```
+
+### Dynamic member lookups
+
+`@dynamicMemberLookup` can resolve method references through key paths, allowing methods to be accessed dynamically without explicit function calls:
+
+```swift
+@dynamicMemberLookup
+struct DynamicKeyPathWrapper<Root> {
+    var root: Root
+
+    subscript<Member>(dynamicMember keyPath: KeyPath<Root, Member>) -> Member {
+        root[keyPath: keyPath]
+    }
+}
+
+let dynamicCalculator = DynamicKeyPathWrapper(root: Calculator())
+let subtract = dynamicCalculator.subtract
+print(subtract(10))
+```
+
+### Effectful value types
+
+Methods annotated with `nonisolated` and `consuming` are supported by this feature. `mutating`, `throwing` and `async` are not supported for any other component type and will similarly not be supported for methods. Keypaths cannot capture method arguments that are not `Hashable`/`Equatable`, so `escaping` is also not supported.
```

Comment:
A few questions:

* How does `consuming` work? (Maybe there's a statement that ought to be here about whether key path roots and/or arguments can be noncopying?)
* `throwing` should probably be spelled `throws`.
* Perhaps instead of singling out `escaping`, you should simply state that keypaths cannot capture closure arguments? (I assume this is the case?)

Reply:
I have updated this section with your suggestions.